Performance improvements down from 38 minutes to 10 minutes for 1000 product imports + other changes

naman03malhotra opened the pull request:
TLDR.
- Product import performance improvements, down from 38 minutes to 10 minutes for 1000 products on Pressable/Atomic Infra.
- I've moved away from a mix of Rest API and GraphQL, to completely fetching records from GraphQL, reducing the time further.
- Added a progress bar for improved user experience.
- Refactored the code to be a bit more readable.
- Added `--verbose` argument for more detailed performance logging.

### Details of the changes

Goal: Significantly improve the performance and user experience of the Shopify-to-WooCommerce product import process via WP-CLI.

Summary of Changes:

**Performance Enhancements (Import Time Reduction):**

- Before: Importing 1000 products took approximately 38 minutes on Pressable/Atomic infrastructure using a mix of REST API and GraphQL calls.
- After: Import time for the same 1000 products is reduced to approximately 10 minutes.

**Technical Details:**

- Pure GraphQL Implementation: Replaced all previous product-fetching REST API calls with Shopify's GraphQL Admin API. This leverages GraphQL's ability to fetch complex nested data (products, variants, images, collections, metafields) in a single request per batch, significantly reducing network latency and complexity.

- Batch Processing with Cursors: Implemented a batch processing mechanism using GraphQL's cursor-based pagination (first, after). This allows fetching products in manageable chunks (--perpage, default 250) without overwhelming the API or the server, and ensures we can resume or process large catalogs efficiently. Earlier we were fetching data using REST APIs and each product would fetch additional metadata using the GraphQL APIs, which would take an additional 0.3 seconds per product.

**Improved User Experience:**

- Progress Bar: Added a WP_CLI\Utils\make_progress_bar to provide real-time feedback on the import progress, showing the estimated number of products processed against the total. The total count is estimated using the REST API's count.json endpoint (as GraphQL doesn't provide a direct count for complex queries).

- Verbose Logging (--verbose): Introduced a --verbose flag to output detailed timing and memory usage information for each product and batch processed. This aids in debugging performance bottlenecks and understanding resource consumption during imports.

**Code Refactoring:**

- Readability & Maintainability: Restructured the Migrator_CLI_Products class for better separation of concerns:
Dedicated methods for argument parsing (parse_and_validate_args).
- Clear separation of batch fetching (fetch_product_batch) and batch processing (process_product_batch).
Isolated logic for handling single products (process_single_product).
Centralized GraphQL query definition.


### Testing instructions


Prep:
1. Setup a Shopify store.
2. Grab the Shopify Access token by [creating a custom app](https://help.shopify.com/en/manual/apps/app-types/custom-apps). Make sure the required scopes for product migration are selected.
3. Import the product using a CSV attached below, this should import 1000 products for you.
[Shopif_1000_products_export_1.csv](https://github.com/user-attachments/files/19887200/Shopif_1000_products_export_1.csv)
4. Create some of your own simple and variable products to test the variations. Add images to variations and other data types such as tags, to fully test the data migration capability.
5. Setup an Jurassic Ninja or a more permanent Pressable site with WooCommerce installed.
6. Make sure you have SSH access.
7. Clone this plugin in the plugins directory and switch to this branch.
8. Activate the plugin from plugins tab.
9. Rename the `config-example.php` to `config.php` and add the key and your Shopify domain.
10. You are all set!

Testing instructions:
1. SSH into your Woo site.
2. Run `wp migrator products` to start the migration.
3. It should start a progress bar and 1000 products should be imported around 10 minutes. If you switch to trunk branch, it should take around 38 minutes.
4. Play around with different options for the command `wp migrator products --limit=100 --perpage=10 --status=active` to make sure the filters are working well.
5. Verify all the data is successfully migrated including images.
6. Try to create a complex product on Shopify with different variations, their individual images, tags etc and make sure they migrate properly.

<img width="1723" alt="Screenshot 2025-04-23 at 4 14 00 PM" src="https://github.com/user-attachments/assets/a76317b3-0d92-4bef-9cb8-c8aab413a3f7" />


Note: That variation support is up to 100 variations per product, we can plan to add support for more in the next iteration.

naman03malhotra committed:
changing ratelimit strategy

naman03malhotra committed:
added perf metrics

naman03malhotra committed:
added perf metrics

naman03malhotra committed:
update porduct limit

naman03malhotra committed:
removed perf logging

naman03malhotra committed:
removed some changes

naman03malhotra committed:
removed dep warning

naman03malhotra committed:
move to GQL

naman03malhotra committed:
move to GQL

naman03malhotra committed:
move to GQL

naman03malhotra committed:
move to GQL

naman03malhotra committed:
move to GQL

naman03malhotra committed:
move to GQL

naman03malhotra committed:
Merge pull request #18 from woocommerce/naman-move-to-gql

Move to gql

naman03malhotra committed:
increased variant capacity

naman03malhotra committed:
Revert "increased variant capacity"

This reverts commit 571d70606cce8f61e296758eef97f1d737917d31.

naman03malhotra committed:
memory log

naman03malhotra committed:
memory log

naman03malhotra committed:
Refactor migrator function

naman03malhotra committed:
Added progress bar

naman03malhotra committed:
Added a flag for verbose output

naman03malhotra committed:
changed messaging

naman03malhotra committed:
changed messaging

naman03malhotra committed:
linting changes

naman03malhotra committed:
added doc block comments

naman03malhotra committed:
added doc block comments

naman03malhotra committed:
improvements around batch fetch

naman03malhotra committed:
improvements around batch fetch

naman03malhotra committed:
update readme and skip update option

naman03malhotra committed:
update readme

naman03malhotra committed:
fetch total product count to respect filters as well

naman03malhotra committed:
update doc blocks

jimjasson reviewed (commented):
Hey @naman03malhotra, 

I tested this PR with a Shopify export of 782 products (173 of which were Variable Products):

[products_export_782.csv](https://github.com/user-attachments/files/19908474/products_export_782.csv)

and confirmed that the import duration dropped from **24 minutes** when using `trunk` to **5 minutes** when using this branch.

I also tested the two paid WooCommerce.com extensions on the same Shopify site and:
- [Migrate & Import Shopify to WooCommerce](https://woocommerce.com/products/migrate-import-shopify-to-woocommerce/) failed after importing 50 products and; 
- [Shopify Migration for WooCommerce](https://woocommerce.com/products/shopify-migration-for-woocommerce/) failed after importing 400 products. 

Here are my initial comments regarding the PR and the code changes: 

### PR format

The PR introduces significant changes and it is not very straight-forward to the reviewer what has changed and why. I would appreciate a more detailed technical spec in the PR description, better analyzing the changes and providing reasoning. 

### Coding standards 

Most of the changes don't follow the WordPress/WooCommerce coding standards -- for example, variables assignments are not aligned. If we plan to bring parts of this extension into core, then we must ensure that we follow core's coding standards to avoid re-doing the work then. 

### Fields mapper 

There isn't a clear indication about which Shopify product fields are matched to which WooCommerce product fields -- I was able to figure out the connections by reading the code, but I think that we should document this mapper somewhere. 

Some Shopify fields are also not mapped to any WooCommerce fields. 

For example, in Shopify Category, Type and Collections are all different ways to define categories with subtle differences -- for example, in the Category field you can only select a category from a pre-defined set of categories that Shopify provides while in the Type field you can type in a custom category. 

![Markup on 2025-04-25 at 15:45:03](https://github.com/user-attachments/assets/a520ae51-aa57-4a13-a812-f44cec04a63d)

In the migrator, we only map the Collections field to WooCommerce categories and we ignore all other fields. Was this perhaps done because this migrator was developed with a specific site in mind? 

Other migrator plugins, provide some options on how to deal with Shopify fields that don't exist in WooCommerce -- see an example from the [Migrate & Import Shopify to WooCommerce](https://woocommerce.com/products/migrate-import-shopify-to-woocommerce/) plugin: 

![Markup on 2025-04-25 at 13:38:21](https://github.com/user-attachments/assets/9ce20bf1-eeb1-4a9e-8871-1be03a6e3226)

I think this requires further product research before making any decisions. Perhaps we could set up a proposed mapper and then review it with James? 

### Building on an existing plugin

As I mentioned above this plugin was developed with a specific store in mind, so some of the implementation decisions might be opinionated. Moreover, this plugin supports additional data for importing (see orders, coupons, etc), while we are trying to focus only on products in this PR and this creates a bit of a confusion in the code.

Here are some examples: 
- The `Migrator_CLI_Products::migrate_products` method starts by calling `Migrator_CLI_Utils::health_check();`, which checks if `wc_get_orders` exists to determine if WooCommerce is active. This seems very specific to the orders import and there are better ways to check if WooCommerce is active. 
- Even though you replaced all REST API calls with GraphQL requests, the word `rest` is still used in variable names, like `rest_id` in `Migrator_CLI_Products::process_single_product`.
- Classes like `/includes/WooPayments/class-migrator-cli-woopayments-customers.php';` need to be manually commented out as they trigger fatal errors.

If we plan to release an MVP only supporting products import, shouldn't all other importers be removed for now? Maybe it makes more sense to work on a fresh repository to make things more clear or move the code from this repo to a legacy repo for reference and then continue building the MVP on this repo?

### Removing hooks 

In `Migrator_CLI_Products::migrate_products`, the `Migrator_CLI_Products::disable_hooks` method runs that disables a lot of hooks that run when a post/product is created, updated and saved. Is this added for performance reasons? Have you checked the implications behind this decision? For example, have you checked if the `product_attributes_lookup` and `product_meta_lookup` tables are populated correctly when these hooks are disabled? Have you considered about plugins hooking into these hooks and adding data when a product is created? 

In my test, if I remove the `Migrator_CLI_Products::disable_hooks`, I only get an additional 50 seconds of import time for 782 products. 

### Warnings

When running the import command on a fresh site, I see the following warnings: 

```
Notice: Function _load_textdomain_just_in_time was called <strong>incorrectly</strong>. Translation loading for the <code>woocommerce</code> domain was triggered too early. This is usually an indicator for some code in the plugin or theme running too early. Translations should be loaded at the <code>init</code> action or later. Please see <a href="https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 6.7.0.) in /srv/htdocs/__wp__/wp-includes/functions.php on line 6121
```

and

```
Deprecated: Creation of dynamic property WP_Object_Cache::$memcache_debug is deprecated in /srv/htdocs/wp-content/plugins/migrator-cli/includes/class-migrator-cli-utils.php on line 269
```

### Logs

All error messages are displayed directly in the console. For large imports with a lot of failures, it might be difficult to figure out what failed and where. Have you considered printing a log with all the errors in the WooCommerce Logs after the import is completed?

naman03malhotra committed:
linting changes

naman03malhotra commented:
>PR format

Thanks for the detailed review @jimjasson! You're right, the PR description could benefit from more technical detail. I have updated the PR description to include:
- A breakdown of the key changes, such as the move to a batch processing model using GraphQL cursors.
- Explanation of why GraphQL was chosen over the previous mixed approach (e.g., reduced network overhead, fetching related data in one go).
- Details on the refactoring, highlighting specific areas that were restructured for better readability and maintainability (e.g., separating fetch/process logic, centralizing argument parsing).
- The rationale for adding the progress bar and --verbose flag."

but let me know if you need more specific info.

>Coding standards

I've addressed some of the lining issues here: https://github.com/woocommerce/migrator-cli/pull/17/commits/23c5e67bae22c4a36009346b9dfd090ca6871b2d

> Field mapper

You are right, a mapping table can be really useful and must be documented, so here is iteration one of the mapping.

Shopify Product Field Mapping to WooCommerce
| Shopify Field (GraphQL Node) | WooCommerce Field / Concept | Notes |
| :------------------------------------------- | :---------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------------------------------- |
| id (Product GID) | _original_product_id (Meta) | Stores the numeric REST ID part of the GID. Used for lookups. |
| title | Product Name (post_title) | Set via WC_Product::set_name(). |
| handle | Product Slug (post_name) | Set via WC_Product::set_slug(). |
| bodyHtml | Product Description (post_content) | HTML is sanitized via sanitize_product_description(). |
| status | Product Status (post_status) | ACTIVE maps to publish, others map to draft. |
| createdAt | Product Creation Date (post_date_gmt, post_date) | Set via WC_Product::set_date_created(). |
| vendor | product_brand Taxonomy Term | Requires product_brand taxonomy to exist. Finds or creates term. |
| tags | product_tag Taxonomy Terms | Finds or creates terms based on tag names. |
| onlineStoreUrl | Catalog Visibility (_visibility term) & _migration_data (Meta) | If null, sets visibility to hidden. If set, stores URL in _migration_data['original_url']. |
| options (name, position, values) | Product Attributes (pa_* taxonomies & terms) | Creates global attributes (pa_*) and terms if they don't exist. Sets them on the product as variation attributes (set_variation(true)). |
| featuredImage (id, url, altText) | Featured Image (_thumbnail_id Meta) | Image is sideloaded. Mapping stored in _migration_data. |
| images (id, url, altText) | Product Gallery (_product_image_gallery Meta) & WP Attachments | Images are sideloaded. Mapping stored in _migration_data. Gallery excludes the featured image. |
| variants (Count) | Product Type (WC_Product_Simple / WC_Product_Variable) | If count(variants) > 1, it's Variable. |
| variants (edges[0]->node - for Simple) | (See individual fields below) | If product is Simple, data from the first variant is used for base product fields. |
| variants->...->id (Variant GID) | Variation ID / _original_variant_id (Meta) | Numeric REST ID stored in meta (on variation, or simple product if only one variant). Mapping GID -> WC ID in _migration_data. |
| variants->...->price | Regular Price (_regular_price Meta) | On variation or simple product. |
| variants->...->compareAtPrice | Sale Price (_sale_price Meta) | On variation or simple product. Set only if compareAtPrice < price. Regular price becomes compareAtPrice if sale is active. |
| variants->...->sku | SKU (_sku Meta) | On variation or simple product. |
| variants->...->inventoryManagement | Manage Stock (_manage_stock Meta) | On variation or simple product. true if value is 'SHOPIFY'. |
| variants->...->inventoryPolicy | Stock Status (_stock_status Meta) | On variation or simple product. outofstock if inventoryQuantity <= 0 and policy is DENY, otherwise instock. |
| variants->...->inventoryQuantity | Stock Quantity (_stock_quantity Meta) | On variation or simple product. |
| variants->...->weight, weightUnit | Weight (_weight Meta) | On variation or simple product. Converted to store's weight unit. |
| variants->...->position | Variation Menu Order (menu_order) | Only for variations. |
| variants->...->image (id, url, altText) | Variation Image (_thumbnail_id Meta) | Only for variations. Uses main image mapping or sideloads directly if needed. |
| variants->...->selectedOptions (name, value) | Variation Attributes (Meta attribute_pa_*) | Only for variations. Maps selected option value slug to the corresponding attribute taxonomy meta key. |
| collections (handle, title) | product_cat Taxonomy Terms | Finds or creates categories based on collection handle/title. |
| metafields (namespace='global', key, value) | _migration_data['metafields'] (Meta) | Stores global metafields in a structured array within the migration data meta. |
| metafields (ns='global', key='title_tag') | Yoast SEO Title (_yoast_wpseo_title Meta) | Requires Yoast SEO plugin. Falls back to product name. |
| metafields (ns='global', key='description_tag') | Yoast SEO Description (_yoast_wpseo_metadesc Meta) | Requires Yoast SEO plugin. Falls back to product description/excerpt. |
Shopify Fields Currently Ignored:
Shopify Standard Category: The predefined category taxonomy from Shopify (e.g., Apparel & Accessories > Clothing > Shirts & Tops). This is not directly available in the main product GraphQL query used and is not mapped.
Product Type: The custom string field productType in Shopify is fetched by the GraphQL query but not mapped to any WooCommerce field (like category, tag, or attribute).




>Building on an existing plugin

>If we plan to release an MVP only supporting products import, shouldn't all other importers be removed for now? Maybe it makes more sense to work on a fresh repository to make things more clear or move the code from this repo to a legacy repo for reference and then continue building the MVP on this repo?

I am open to create a new repo, and I agree that we should remove the other plugin code, which I plan to do in a separate PR, however this point came up during a sync call we agreed that creating a new repo might not offer the value offsetting the maintaining of multiple legacy codebases. But if we both find a consensus, then I'm fully down with the idea of creating a new repo.

Consider a +1 from my end, but let's think it through.


>Removing hooks

Yes, the intention behind disabling these hooks was purely for performance optimization during the bulk import, to minimize the overhead on each product save. 

>In my test, if I remove the Migrator_CLI_Products::disable_hooks, I only get an additional 50 seconds of import time for 782 products.

That might seem like a small amount of time, but I am afraid it can quickly add up for a large quantity of products.

> For example, have you checked if the product_attributes_lookup and product_meta_lookup tables are populated correctly when these hooks are disabled?

I haven't tested this, but I believe the table data for these can be generated later using (WooCommerce > Status > Tools) and doesn't have to be part of the import process. We could consider disabling it by default, with an option to enable it for users who want the lookup generation process. What do you think?

>Warnings

Thanks for pointing it out, I'll look into the warnings.


>Logs

That's would be good addition, but do we need to for MVP? I was under the impression that we'll add things incrementally. I am open to it if we feel this should be a part of MVP as well.


----

I really appreciate all these suggestions and I am open to implement most of them, however if we are going for with a MVP route, and build incrementally I think we can phase out features. We need to decide if we need to prioritize launch speed vs features/additons. We should plan this out with James. Also I left a question about creating new repo.

naman03malhotra committed:
linting changes

jimjasson commented:
>  I have updated the PR description

It's way easier now to follow the code changes -- thanks for that! 💯

> I've addressed some of the lining issues here

Great!

> You are right, a mapping table can be really useful and must be documented, so here is iteration one of the mapping.

That's great! 

This list needs to be reviewed by James and, if it looks good to him, we should add it to some kind of plugin documentation, so we can refer back to it. 

I am a bit concerned that some 'Category' fields are left out and I guess that this decision was based on the specific needs of the site for which this plugin was specifically developed for. Let's share this with James for approval. 

> I am open to create a new repo

Currently under discussion -- let's circle back to this later. 

> I haven't tested this, but I believe the table data for these can be generated later using (WooCommerce > Status > Tools) and doesn't have to be part of the import process. We could consider disabling it by default, with an option to enable it for users who want the lookup generation process. What do you think?

I would echo the following statement you made:

> That's would be good addition, but do we need to for MVP?

The migrator is already close to 4 times faster than it was, even without removing the hooks. It is also worth noting that since there isn't any way to migrate data from a Shopify app to a WooCommerce extension, I imagine that the most common use case is to migrate Shopify to a fresh store, without any or many plugins using these hooks and thus running this hooks having a big impact on performance.

I am more in favor of adding a command option to skip hooks, but leave them running by default.

naman03malhotra committed:
review changes

jimjasson commented on includes/class-migrator-cli-utils.php:
```diff
@@ -101,20 +101,115 @@ public static function rest_request( $endpoint, $body = array() ) {
+
+		while ( $retry_count <= $max_retries ) {
+			$response = wp_remote_post(
+				'https://' . SHOPIFY_DOMAIN . '/admin/api/2023-04/graphql.json',
```
@naman03malhotra This API version doesn't seem to be supported any longer: https://shopify.dev/docs/api/release-notes/previous-versions/2023-04.

Have you tried using the latest API version here?

naman03malhotra committed:
update the api to latest version + other changes

naman03malhotra committed:
update the api to latest version + other changes

naman03malhotra committed:
update the api to latest version + other changes

naman03malhotra committed:
update default limit to optimize query cost

naman03malhotra committed:
remove unwanted changes

naman03malhotra committed:
increased variant capacity

naman03malhotra committed:
increased variant capacity

naman03malhotra committed:
added flat to enable/disable hooks

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor and cleanup

naman03malhotra committed:
big refactor 2

naman03malhotra committed:
big refactor 2

naman03malhotra committed:
big refactor 2

naman03malhotra commented:
closing it in favor of: https://github.com/woocommerce/migrate-to-woo/pull/1
